ruby/phase-1.md

```diff
@@ -0,0 +1,575 @@
+`Client#workflow_handle` method. Here the interface of the handle:
+
+```ruby
+class Temporal::WorkflowHandle
+  # Return the result of the workflow execution or raise if a workflow is still running after
+  #   waiting for for up to :timeout seconds.
+  # Passing timeout: nil will not wait if the working is still running
+  def result(timeout: Integer) -> Any
+
+  # Returns more information about a workflow
+  def describe() -> Temporal::WorkflowExecutionInfo
+
+  # Sends a cancellation request
+  def cancel() -> void
+
+  # Sends a query and awaits for the response
+  def query(name, *args) -> Any
+
+  # Sends a signal
+  def signal(name, *args) -> void
+
+  # Terminates a workflow
+  def terminate(reason, details: nil) -> void
+end
+```
+
+### Alternatives considered
+
+We thought about configuring the `Temporal::Client` with a namespace, similar to the Python SDK.
+It does bring the benefit of avoiding namespace to subsequent calls to the client. However we feel
+like it is an unjustified and artificial limitation that will impact some multi-tenant applications.
+Besides this can always be brought in the shape of a `NamespacedClient` if there's a need for it.
+
+
+## Worker
+
+The worker interface allows SDK users to start processing registered workflow and activities. These
+will then be polled for on the specified namespaces & task queues and executed.
+
+```ruby
+config = Temporal::Configuration.new(url: 'localhost:7233')
+
+worker_1 = Temporal::Worker.new(config, 'my-namespace', 'my-task-queue')
+worker_1.register_workflow(MyWorkflow) # uses 'MyWorkflow' as a name
+worker_1.register_activity(MyActivity) # uses 'MyActivity' as a name
+worker_1.start
+
+worker_2 = Temporal::Worker.new(config, 'my-namespace', 'another-task-queue')
+worker_2.register_workflow(MyWorkflow, name: 'my-workflow') # uses 'my-workflow' as a name
+worker_2.register_activity(MyActivity, name: 'my-activity') # uses 'my-activity' as a name
+worker_2.start
+
+# later
+worker_1.stop
+worker_2.stop
+```
+
+Notes:
+
+- This is not complete example of all configuration options for a worker
+- A worker will use the workflow/activity class name as a name unless it is explicitly provided (
+  `worker_2` demonstrated this approach)
+- Each worker will run in its own thread
+- A blocking `Worker#run` method will be available for convenience of running a single worker in
+  a process until SIGTERM or SIGINT is received
+- In case connection sharing is a concern we can create and memoize a connection on the instance of
+  Temporal::Configuration
+
+And here is the `Worker` interface:
+
+```ruby
+class Worker
+  # Create a new instance of a Worker
+  def self.new(config: Temporal::Configuration, namespace: String, task_queue: String, **options) -> Worker
+
+  # Register a workflow with an optional :name
+  def register_workflow(Class, name: nil) -> void
+
+  # Register an activity with an optional :name
+  def register_activity(Class, name: nil) -> void
```

Comment:
There is a bit of an issue we realized in Python wrt naming. Workflow and activity implementations needs to carry their names because people use those implementation references to perform invocation. If I registered an activity here with a different name and then invoked `execute_activity(MyActivity)` inside a workflow, things would break. And that form of invocation is encouraged over using strings in our SDKs.

Is there a preferred Ruby pattern for adding metadata to classes or functions?

Reply:
I made the exact same comments above

Reply:
Another case is replacing an existing activity with an implementation in another language without changing the caller. I just don't see why wouldn't we allow for this sort of flexibility

Reply:
> The main purpose is being able to register an activity/workflow twice.

This isn't a real concern and definitely not something that should affect our API. A user can wrap or delegate to another activity if they want a different name. It hasn't been a concern in any of our SDKs up until now and I don't think it ever will be.

Reply:
> I just don't see why wouldn't we allow for this sort of flexibility

Having two places to customize a name is ambiguous